File: prototype/rails/src/app/domain/escrow/payout_period.rb
require "date"

module Domain
  module Escrow
    # The Monday-to-Sunday week a payout run settles: the most recently
    # completed one as of a given moment.
    PayoutPeriod = Data.define(:first_day, :last_day) do
      def self.ending_before(as_of)
        date = as_of.to_date
        first_day = date - ((date.wday - 1) % 7) - 7

        new(first_day: first_day, last_day: first_day + 6)
      end

      # Everything dated at or before this instant belongs to the period, which
      # is what makes a second run of the same period a no-op.
      def ends_at
        Time.utc(last_day.year, last_day.month, last_day.day, 23, 59, 59)
      end

      def covers?(moment)
        (first_day..last_day).cover?(moment.to_date)
      end

      def label
        "#{first_day} to #{last_day}"
      end
    end
  end
end

File: prototype/node/src/app/core/escrow/payout-period.ts
/**
 * The Monday-to-Sunday week a payout run settles: the most recently
 * completed one as of a given moment.
 */
export type PayoutPeriod = { firstDay: string; lastDay: string }

function toUtcDateString(date: Date): string {
  return date.toISOString().slice(0, 10)
}

function shiftUtcDays(date: Date, days: number): Date {
  const shifted = new Date(date.getTime())
  shifted.setUTCDate(shifted.getUTCDate() + days)
  return shifted
}

export function payoutPeriodEndingBefore(asOf: Date): PayoutPeriod {
  const daysSinceMonday = (((asOf.getUTCDay() - 1) % 7) + 7) % 7
  const firstDay = shiftUtcDays(asOf, -(daysSinceMonday + 7))
  const lastDay = shiftUtcDays(firstDay, 6)
  return { firstDay: toUtcDateString(firstDay), lastDay: toUtcDateString(lastDay) }
}

// Everything dated at or before this instant belongs to the period, which is
// what makes a second run of the same period a no-op.
export function payoutPeriodEndsAt(period: PayoutPeriod): Date {
  return new Date(`${period.lastDay}T23:59:59.999Z`)
}

export function payoutPeriodCovers(period: PayoutPeriod, moment: Date): boolean {
  const day = toUtcDateString(moment)
  return day >= period.firstDay && day <= period.lastDay
}

export function payoutPeriodLabel(period: PayoutPeriod): string {
  return `${period.firstDay} to ${period.lastDay}`
}
